src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/UnusedTrackerWithR8.groovy

```diff
@@ -0,0 +1,276 @@
+        R8Command.Builder builder = R8Command.builder()
+
+        populateBuilderWithProjectFiles(builder)
+
+        // add all dependencies as library jars to avoid warnings
+        for (File dep : dependencies) {
+            Path path = Paths.get(dep.getAbsolutePath())
+            builder.addLibraryFiles(path)
+        }
+
+        // disable everything, we just want to get a list of
+        // all project classes.
+        List<String> configs = new ArrayList<>()
+        configs.add("-dontshrink")
+        configs.add("-dontoptimize")
+        configs.add("-dontobfuscate")
+        configs.add("-ignorewarnings")
+        configs.add("-dontwarn")
+
+        builder.addProguardConfiguration(configs, Origin.unknown())
+
+        // Set compilation mode to debug to disable any code of instruction
+        // level optimizations.
+        builder.setMode(CompilationMode.DEBUG)
+
+        final List<String> keepRules = new ArrayList<>()
+
+        builder.setProgramConsumer(new ClassFileConsumer() {
+            @Override
+            void accept(ByteDataView byteDataView, String s, DiagnosticsHandler diagnosticsHandler) {
+                String name = typeNameToExternalClassName(s)
+                keepRules.add("-keep,includedescriptorclasses class " + name + " { *; }")
+            }
+
+            @Override
+            void finished(DiagnosticsHandler diagnosticsHandler) {}
+        })
+
+        R8.run(builder.build())
+        return keepRules
+    }
+
+    @CompileStatic
+    void populateBuilderWithProjectFiles(BaseCommand.Builder builder) {
+        addJDKLibrary(builder)
+
+        for (File f : projectFiles) {
+            Path path = Paths.get(f.getAbsolutePath())
+
+            if (f.getAbsolutePath().endsWith(".class")) {
+                byte[] bytes = Files.readAllBytes(Paths.get(f.getAbsolutePath()))
+                builder.addClassProgramData(bytes, Origin.unknown())
+            } else {
+                builder.addProgramFiles(path)
+            }
+        }
+    }
+
+    @CompileStatic
+    private void collectClassFiles(Path dir, Collection<File> result) {
+        File file = dir.toFile()
+        if (file.exists()) {
+            File[] files = file.listFiles()
+            if (files != null) {
+                for (File child : files) {
+                    if (child.isDirectory()) {
+                        collectClassFiles(child.toPath(), result)
+                    } else {
+                        Path relative = child.toPath()
+                        if (isClassFile(relative)) {
+                            result.add(child.getAbsoluteFile())
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @CompileStatic
+    static void addJDKLibrary(BaseCommand.Builder builder) {
+        String JAVA_HOME = System.getenv("JAVA_HOME")
+        builder.addLibraryResourceProvider(JdkClassFileProvider.fromJdkHome(Paths.get(JAVA_HOME)))
```

Comment:
I'm not sure we can/should rely on this. Perhaps this needs to be exposed to the user directly? We could maybe figure out a default setting based on the configuration of the `compileMain` task or somethign.